src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -325,71 +326,78 @@ async delegate
 
                         lock (SyncObject)
                         {
+                            Verify.NotDisposed(this);
+
                             DependenciesSnapshotProvider.SnapshotChanged += OnDependenciesSnapshotChanged;
 
-                            Verify.NotDisposed(this);
                             Task<IProjectVersionedValue<IProjectTreeSnapshot>> nowait = SubmitTreeUpdateAsync(
                                 (treeSnapshot, configuredProjectExports, cancellationToken) =>
                                 {
-                                    IProjectTree dependenciesNode = CreateDependenciesFolder(null);
+                                    IProjectTree dependenciesNode = CreateDependenciesFolder();
 
                                     // TODO create providers nodes that can be visible when empty
                                     //dependenciesNode = CreateOrUpdateSubTreeProviderNodes(dependenciesNode, 
                                     //                                                      cancellationToken);
 
-                                    return Task.FromResult(new TreeUpdateResult(dependenciesNode, true));
-                                });
+                                    return Task.FromResult(new TreeUpdateResult(dependenciesNode, lazyFill: true));
+                                },
+                                GetNextTreeUpdateCancellationToken());
                         }
 
                     },
                     registerFaultHandler: true);
             }
-        }
 
-        private Task OnUnconfiguredProjectUnloading(object sender, EventArgs args)
-        {
-            UnconfiguredProject.ProjectUnloading -= OnUnconfiguredProjectUnloading;
-            DependenciesSnapshotProvider.SnapshotChanged -= OnDependenciesSnapshotChanged;
+            IProjectTree CreateDependenciesFolder()
```

Comment:
>IProjectTree CreateDependenciesFolder() [](start = 12, length = 39)

I don't know what discussion you guys had over breakfast, but these local functioned code is harder to read .... :)

Reply:
What exactly is the new convention and what are benefits?

---
In reply to: [226698249](https://github.com/dotnet/project-system/pull/4143#discussion_r226698249) [](ancestors = 226698249)

Reply:
Readability is pretty subjective, so let me explain how I see the benefit.

The more top level members a type has, the more potential dependencies there are between them. I find this makes it harder to reason about invariants and usage of members.

By nesting members within other members, it communicates (enforces) separation. These aggregate members are easier to understand in isolation. You can also see the domain of values exchanged and other invariants, which often highlights opportunities to simplify things, IME.

Complexity with local functions can come about when you capture and mutate many variables. There's talk of adding `static` local functions that can't capture, for this reason. For now it's a good idea to avoid capturing in complex ways, especially if mutating.

Introducing local functions where none previously existed can increase readability. Function names are documentation.

This style of nesting is familiar to me from work in other languages where I grew to really like it. I encourage folks to give it a go. I'm doing a bunch of work on the dependency node at the moment and will limit such changes to that area of the code for now.

Reply:
still confusing.. separating code to a private function is done to reuse some functionality and it is fine if other class members call it. Yes if it is not done right then it could lead to some harder to follow logic that might change the state of the object. But that can b e said about anything.. 

On the other side to me if the method has 20+ lines it is already too complicated and local functions making it a little harder to read. Anyway, it is subjective as you said .. 

---
In reply to: [226707589](https://github.com/dotnet/project-system/pull/4143#discussion_r226707589) [](ancestors = 226707589)

Reply:
i will try it out and will see if i get used to it :)

---
In reply to: [226717284](https://github.com/dotnet/project-system/pull/4143#discussion_r226717284) [](ancestors = 226717284,226707589)

Reply:
Oh, one more thing about local functions&mdash;it sometimes means you can push state "down" in scope, from field to local.

I guess the overall benefit I see is restricting the scope of things.

If a local function becomes useful in another scenario, it can be lifted out.

A pattern I like (and I think it's used in Roslyn) is to make sure all local functions are at the end of the method, and if the method has `void` return type, to include a redundant `return;` before the first local function, just to make it clear that the top level method is finished.